onnxruntime/core/optimizer/propagate_cast_ops.cc

```diff
@@ -121,18 +118,41 @@ static bool IsRelevantOutput(const Node* node, const NodeArg* output) {
 
   auto it = opcode_to_output_map.find(node->OpType());
   if (it != opcode_to_output_map.cend()) {
-    const std::vector<int>& selected_outputs = it->second;
+    const auto& selected_outputs = it->second;
     int input_index = optimizer_utils::IndexOfNodeOutput(*node, *output);
     return std::find(selected_outputs.begin(), selected_outputs.end(), input_index) != selected_outputs.end();
   }
   return true;
 }
 
+namespace {
+// borrowed from providers/common.h
+template <class AssociativeContainer, class Key>
+inline bool Contains(const AssociativeContainer& container, const Key& key) {
+  return container.find(key) != container.end();
+}
+}  // namespace
+
 // Check whether the given opcode is fp16 allowed for the given level of optimization.
-static bool IsFP16Allow(const std::string& op_type, size_t level, const FP16AllowOps& fp16_allow_ops) {
-  bool fp16_allow = false;
-  for (size_t i = 0; i <= level && i < fp16_allow_ops.size() && !fp16_allow; ++i) {
-    fp16_allow = fp16_allow_ops[i].find(op_type) != fp16_allow_ops[i].end();
+static bool IsFP16Allow(const std::string& op_type, size_t level, const FP16AllowOps& fp16_allow_level0_ops) {
```

Comment:
This modified code is not very readable and maintainable.

Reply:
Please, provide a specific suggestion to make this change more readable.

Reply:
The changes here not according the design. TranformerLevel::Level2 is not related at all to cast propagation level and Level0, Level1, etc. are not in one-to-one relation with the fp16_allow_sets. This error prone as there is no gurantee that enums follow the ordinal numbers, Level1==1 and Level2==2 other than coding decipline. We could create a new set of fp16_allow_sets and/or re-arrange the existing opcodes in to more fp16_allow_sets. Please revert changes to the functions and code organization.
I don't know what problem this code change is trying to fix!

Reply:
The error could result in segmentation error and core dump if Level2 happens to be more than 2.

Reply:
Thank you for pointing that out. So would initializing `MaxSupportedOptimizationLevel` = 2 take care of the problem?

Reply:
The problem this code is fixing is unncessary copying and memory allocation of the const static data that can be used as is and not copied to every instance of the transformer.

Reply:
The problem with hardening code this way makes it prone to error and difficult to change, for exmple adding a level requires changing different things correctly. Hardening does not improve training throughput as graph tranformation is only performed at the beginning not at eatch training step.